_user/voxeldemo/main.c

```diff
@@ -0,0 +1,371 @@
+			*pixels += (((unsigned)b0 << 0) & 0x0000ff);
+
+			pixels++;
+		}
+	}
+}
+
+
+void voxel_genLand(voxel_t *v)
+{
+	unsigned int tmp;
+	int p, i, j, k, k2, p2;
+	int a, b, c, d;
+
+	/* Start from a plasma clouds fractal */
+	v->mapHeight[0] = 128;
+	for (p = 1024; p > 1; p = p2) {
+		p2 = p >> 1;
+		k = p * 8 + 20;
+		k2 = k >> 1;
+		for (i = 0; i < 1024; i += p) {
+			for (j = 0; j < 1024; j += p) {
+				a = v->mapHeight[(i << 10) + j];
+				b = v->mapHeight[(((i + p) & 0x3ff) << 10) + j];
+				c = v->mapHeight[(i << 10) + ((j + p) & 0x3ff)];
+				d = v->mapHeight[(((i + p) & 0x3ff) << 10) + ((j + p) & 0x3ff)];
+
+				v->mapHeight[(i << 10) + ((j + p2) & 0x3ff)] =
+					clamp(((a + c) >> 1) + (rand() % k - k2));
+				v->mapHeight[(((i + p2) & 0x3ff) << 10) + ((j + p2) & 0x3ff)] =
+					clamp(((a + b + c + d) >> 2) + (rand() % k - k2));
+				v->mapHeight[(((i + p2) & 0x3ff) << 10) + j] =
+					clamp(((a + b) >> 1) + (rand() % k - k2));
+			}
+		}
+	}
+
+	/* clamp unwanted noise level */
+	for (i = 0; i < 1024 * 1024; i += 1024) {
+		for (j = 0; j < 1024; j++) {
+			if (v->mapHeight[i + j] < 50)
+				v->mapHeight[i + j] = 50;
+
+			v->mapHeight[i + j] -= 50;
+		}
+	}
+
+	/* Initialize color map */
+	for (i = 0; i < 1024 * 1024; i += 1024)
+		for (j = 0; j < 1024; j++)
+			v->mapColor[i + j] = v->mapHeight[i + j];
+
+	/* clamp water level */
+	for (i = 0; i < 1024 * 1024; i += 1024) {
+		for (j = 0; j < 1024; j++) {
+			if (v->mapHeight[i + j] < 100)
+				v->mapHeight[i + j] = 100;
+
+			v->mapHeight[i + j] -= 100;
+		}
+	}
+	/* Smooth height map */
+	for (k = 0; k < 20; k++) {
+		for (i = 0; i < 1024 * 1024; i += 1024)
+			for (j = 0; j < 1024; j++) {
+				tmp = v->mapHeight[((i + 1024) & 0xffc00) + j];
+				tmp += v->mapHeight[i + ((j + 1) & 0x3ff)];
+				tmp += v->mapHeight[((i - 1024) & 0xffc00) + j];
+				tmp += v->mapHeight[i + ((j - 1) & 0x3ff)];
+				v->mapHeight[i + j] = tmp >> 2;
+			}
+	}
+
+	/* Smooth color map */
+	for (k = 0; k < 1; k++) {
+		for (i = 0; i < 1024 * 1024; i += 1024)
+			for (j = 0; j < 1024; j++) {
+				tmp = v->mapColor[((i + 1024) & 0xffc00) + j];
+				tmp += v->mapColor[i + ((j + 1) & 0x3ff)];
+				tmp += v->mapColor[((i - 1024) & 0xffc00) + j];
+				tmp += v->mapColor[i + ((j - 1) & 0x3ff)];
+				v->mapColor[i + j] = tmp >> 2;
+			}
+	}
+}
+
+
+void voxel_genPalette(voxel_t *v)
+{
+	unsigned int i, idx = 0;
+	uint8_t step[] = { 50, 50, 50, 50, 10, 10, 10, 15 };
+	uint32_t colors[] = {
+		0x3c789c, 0x4884a8, /* water */
+		0x3d7028, 0x846c4c, /* grass */
+		0xc6b3a3, 0xb8ada2, /* sand 1 */
+		0x846c4c, 0x8c7454, /* sand 2 */
+		0xbcc8d8,           /* frost-snow */
+	};
+
+	for (i = 0; i < sizeof(step); i++)
+		idx += interpolate(&v->palette[idx], colors[i], colors[i + 1], step[i]);
+}
+
+
+int voxel_demo(graph_t *g)
+{
+	unsigned int t = 0;
+	voxel_t v = {
+		.width = g->width,
+		.height = g->height,
+		.pixels = g->data,
+		.cam = { .h = 300, .horiz = 200, .dist = 300 },
+	};
```

Comment:
```suggestion
		.cam = { .h = 300, .horiz = 200, .dist = 300 },
		.bufBack = NULL;
		.mapHeight = NULL;
		.mapColor = NULL;
		.palette = NULL;			
	};
```

Reply:
It is not necessary. Any structure element in a C initializer that is not listed defaults to `0` / `NULL`.